### DISCUSSION TITLE
How do I get better floating point utilisation with this kernel?

### DISCUSSION THREAD

**Discussion by torrance** [ROCm/ROCm#3841] [Q&A]
I've been benchmarking a kernel on both an AMD MI250X versus an NVIDIA A100. I can easily get 99+% (14.9 TFLOPS)  floating point utilisation with the A100, but I'm down at 61.7% (13.4 TFLOPS) with the MI250X. This means that even though the AMD accelerator is capable of faster results, the NVIDIA device currently runs faster.

For reference, here is the roofline plot on the MI250X:

![roofline-kernel](https://github.com/user-attachments/assets/8b673069-49d2-494b-a9f2-17550a40c9e3)

I'd love some help if you can see how to improve floating point utilisation of this kernel.

The kernel is rather big but I have a very simple toy example that captures a fair bit of the work that the kernel does:

```c++
__global__ __launch_bounds__(nthreads)
void kernel(thrust::complex<float>* subgrid, thrust::complex<float>* data) {
    __shared__ thrust::complex<float> cache[nthreads];

    thrust::complex<float> x[threadcoarsen] {};
    for (uint32_t i {}; i < datasz; i += nthreads) {
        __syncthreads();
        cache[threadIdx.x] = data[i + threadIdx.x];
        __syncthreads();

        for (uint32_t j {}; j < nthreads; ++j) {
            const auto datum = cache[j];
            for (uint32_t k {}; k < threadcoarsen; ++k) {
                // Equivalent to x[k] += datum * x[k] but with fma instructions
                // x[k] += datum * x[k];
                cmac(x[k], datum, x[k]);
            }
        }
    }

    uint32_t idx = (blockIdx.y * subgridsz) + threadcoarsen * (threadIdx.x + blockIdx.x * blockDim.x);
    for (uint32_t i {}; i < threadcoarsen; ++i) subgrid[idx + i] = x[i];
}
```

Essentially for each subgrid element, we do a reduction over the data. Ignore the redundancy of the calculation for now - this is just a toy example, and the real kernel does a unique calculation per subgrid element.

The kernel is distributed along a y-dimension that each process a different subgrid, and then an x-dimension for each pixel in the 128x128 px subgrid. In the real kernel, the length of data varies for each subgrid.

I'm already using a lot of standard optimizations: shared memory, coalesced reads and writes, thread coarsening, and 100% FMA instructions (that little `cmac` function does a fused multiply accumulate with complex numbers). I'm not sure what else to do!

<details><summary>Here's the full code if anyone wants to have a go at optimisation:</summary>
<p>

```c++
#include <chrono>
#include <iostream>

#include <hip/hip_runtime.h>
#include <thrust/complex.h>

#define HIPCHECK(res) { hipcheck(res, __FILE__, __LINE__); }

inline void hipcheck(hipError_t res, const char* file, int line) {
    if (res != hipSuccess) {
        throw std::runtime_error(
            "Fatal hipError"
        );
    }
}

/**
 * Complex-valued fused multiply accumulate
 * x += y * z
 */
template <typename T>
__device__ __inline__
void cmac(thrust::complex<T>& x, const thrust::complex<T> y, const thrust::complex<T> z) {
    x.real( fma(y.real(), z.real(), x.real()) );
    x.imag( fma(y.real(), z.imag(), x.imag()) );
    x.real( fma(-y.imag(), z.imag(), x.real()) );
    x.imag( fma(y.imag(), z.real(), x.imag()) );
}

// Kernel configuration constants
const uint32_t nsubgrids {1024}; // we have a stack of 1024 subgrids, each being 128px x 128px
const uint32_t subgridsz {128 * 128};
const uint32_t datasz {1024 * 1024}; // in the real kernel, datasz varies between each subgrid
const uint32_t threadcoarsen {4};
const uint32_t nthreads {128};
const uint32_t nblocks {subgridsz / (nthreads * threadcoarsen)};
static_assert(nblocks * nthreads * threadcoarsen == subgridsz);

__global__ __launch_bounds__(nthreads)
void kernel(thrust::complex<float>* subgrid, thrust::complex<float>* data) {
    __shared__ thrust::complex<float> cache[nthreads];

    thrust::complex<float> x[threadcoarsen] {};
    for (uint32_t i {}; i < datasz; i += nthreads) {
        __syncthreads();
        cache[threadIdx.x] = data[i + threadIdx.x];
        __syncthreads();

        for (uint32_t j {}; j < nthreads; ++j) {
            const auto datum = cache[j];
            for (uint32_t k {}; k < threadcoarsen; ++k) {
                // Equivalent to x[k] += datum * x[k] but with fma instructions
                // x[k] += datum * x[k];
                cmac(x[k], datum, x[k]);
            }
        }
    }

    uint32_t idx = (blockIdx.y * subgridsz) + threadcoarsen * (threadIdx.x + blockIdx.x * blockDim.x);
    for (uint32_t i {}; i < threadcoarsen; ++i) subgrid[idx + i] = x[i];
}

int main() {
    thrust::complex<float>* data {};
    thrust::complex<float>* subgrid {};

    HIPCHECK( hipMalloc(&data, sizeof(thrust::complex<float>) * datasz) );
    HIPCHECK( hipMalloc(&subgrid, sizeof(thrust::complex<float>) * subgridsz * nsubgrids) );
    HIPCHECK( hipDeviceSynchronize() );

    auto start = std::chrono::steady_clock::now();

    hipLaunchKernelGGL(
        kernel, dim3(nblocks, nsubgrids), dim3(nthreads, 1), 0, hipStreamPerThread, subgrid, data
    );

    std::cout << "Waiting for kernel..." << std::endl;
    HIPCHECK( hipStreamSynchronize(hipStreamPerThread) );

    auto stop = std::chrono::steady_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(stop - start);

    std::cout << "Complete (" << duration.count() << " ms)" << std::endl;

    return 0;
}
```

</p>
</details>

I would love any pointers or advice to make full use of the AMD accelerator.

**Comment by jamesxu2** [comment]
Hello @torrance, thanks for the question.

I've profiled your kernel (thanks for providing the source!) and looked at the generated assembly - nothing stands out to me as being obviously wrong. The runtime of your program is dominated by the kernel so there are no obviously extraneous transfers or synchronization. 
Beyond that, your question about optimizing a specific kernel is a bit beyond the scope of what our support team typically handles. I can however provide some general guidance:

- In case you haven't read it, there is a page on performance optimization with HIP in our [docs](https://rocm.docs.amd.com/projects/HIP/en/latest/how-to/performance_guidelines.html#device-memory-access).
- If you have access to the MI250 system hardware, you may consider reviewing our [system setup guide for MI250](https://rocm.docs.amd.com/en/latest/how-to/system-optimization/mi200.html).
- You might consider the tradeoffs of using [lower precision datatypes](https://rocm.docs.amd.com/projects/HIP/en/latest/how-to/performance_guidelines.html#arithmetic-instructions) (like fp16) or risky [compiler optimizations](https://rocm.docs.amd.com/en/docs-5.2.3/reference/rocmcc/rocmcc.html#compiler-reference-guide) like -ffast-math . However, whether that is appropriate requires some domain expertise which we do not have. Regardless, make sure to validate your results with a CPU calculation to make sure you're not creating some unexpected data races.
